### PR TITLE
[Neural Speed] Cont Batching in Offline and Server (GPT-J & Beam Search First)

### PR DESCRIPTION
**Pull request by zhentaoyu**
## Type of Change

feature or bug fix or documentation or others
As title

## Description

detail description 
Jira: [829](https://jira.devtools.intel.com/browse/NLPTOOLKIU-829) && [1152](https://jira.devtools.intel.com/browse/NLPTOOLKIU-1152)

- How to use?
   python usage please see [here](https://github.com/intel/neural-speed/blob/cont_batching/scripts/python_api_example_for_model_server.py)
- Do other models support cont-batching?
   we will support cont-batching in other models in the future and this cont-batching will be the main batching inference way.
- Does it improve performance
   It should be since it saves padding dealing time in offline scenario and makes inferences on the fly in server scenario. we will update GPT-J MlPerf results soon.
- Will it affect logits?
   We think it would not (for-loop `rope` & `kv-concat` & `self-attn`). However, we found it seems to have some different logits (but not too much). It can pass MLPerf offline and server scenarios accuracy tests (int4-int8-per-channel & int4-bf16-per-channel). Will raise another PR if it's a bug in the future. 

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Comment by zhentaoyu** [comment]
gpt-j-6b ext test: https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/43/

File: neural_speed/__init__.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 from transformers import AutoConfig, AutoTokenizer
 
 model_maps = {"gpt_neox": "gptneox", "gpt_bigcode": "starcoder"}
+max_request_num_default = 8
```
Comment (airMeng) [inline neural_speed/__init__.py:24]: source?

File: neural_speed/application/main_pybind.cpp [hunk 2]
```diff
@@ -54,6 +56,241 @@
...
+    assert(token_ids.ndim() == 1 || (token_ids.ndim() == 2 && token_ids.shape(0) == 1));
+  }
+
+  std::string to_string() const {
+    const std::string repr_ids(py::str(py::array_t<int, py::array::c_style>(token_ids.size(), token_ids.data())));
+    return std::to_string(id) + ": " + repr_ids;
+  }
+};
+
+using Response = Query;
```
Comment (airMeng) [inline neural_speed/application/main_pybind.cpp:75]: any specific reasons?

File: neural_speed/application/main_pybind.cpp [hunk 3]
```diff
@@ -54,6 +56,241 @@
...
+  const ResponseCallback response;
+  std::vector<Query> waiting;
+  std::mutex queue_mtx;
+  bool running;
+  gpt_params params;
+  std::string policy;
+  bool scheduler_empty;
+  uint64_t working_size;
+  bool return_prompt;
+  std::thread worker;
```
Comment (airMeng) [inline neural_speed/application/main_pybind.cpp:291]: some comments here might be helpful

File: neural_speed/application/main_pybind.cpp [hunk 4]
```diff
@@ -54,6 +56,241 @@
...
+
+class ModelServer {
+ public:
+  ModelServer(const ResponseCallback& response, const std::string& model_path, bool return_prompt, int max_new_tokens,
+              int n_batch, int ctx_size, int seed, int threads, float repetition_penalty, int num_beams, bool do_sample,
+              int top_k, float top_p, float temperature, int min_new_tokens, float length_penalty, bool early_stopping,
+              int n_keep, int n_discard, bool shift_roped_k, int batch_size, model_vocab::id pad_token,
+              const std::string& memory_dtype, const bool& continuous_batching, const int& max_request_num,
+              const float& model_scratch_enlarge_scale, const std::string& policy, const bool& print_log,
+              const std::function<void()>& init_cb)
```
Comment (airMeng) [inline neural_speed/application/main_pybind.cpp:142]: usually we don't encourage so much arguments, you can pack them into several structs instead

File: neural_speed/models/model_utils/model_types.h [hunk 5]
```diff
@@ -61,6 +61,8 @@
 #define MODEL_SESSION_MAGIC MODEL_FILE_MAGIC_GGSN
 #define MODEL_SESSION_VERSION 1
 
+#define MODEL_MAX_REQUEST_NUM 8
```
Comment (airMeng) [inline neural_speed/models/model_utils/model_types.h:64]: same as python side?

File: neural_speed/models/model_utils/pool.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+bool fcfs_pool::pop(sequence* seq) {
+  if (empty()) {
+    fprintf(stderr, "%s: pool is empty.\n", __func__);
+    return false;
+  }
+  *seq = context.front();
+  context.pop();
+  return true;
+}
```
Comment (airMeng) [inline neural_speed/models/model_utils/pool.cpp:40]: why must return a bool instead of void function?

File: neural_speed/models/model_utils/scheduler.h [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+//  limitations under the License.
+
+#ifndef SCHEDULER_H
+#define SCHEDULER_H
+
+#include "models/model_utils/pool.h"
+#include "models/model_utils/model_utils.h"
+
+// iteration-level worker
+class il_worker {
```
Comment (airMeng) [inline neural_speed/models/model_utils/scheduler.h:22]: ```il``` seems confusing

**Review by airMeng — APPROVED** [review]
LGTM except some minor comments

**Comment by airMeng** [comment]
> gpt-j-6b ext test: https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/43/

no cont_batching in extension test, so we will test cont_batching though python API from now on?

**Comment by zhentaoyu** [comment]
> > gpt-j-6b ext test: https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/43/
> 
> no cont_batching in extension test, so we will test cont_batching though python API from now on?

yes. Added cont-batching offline generation result test in the `test` folder. As for offline perf and `ModelServer`, I will discuss with @DDEle @zhenwei-intel how to add a test and how to measure its perf after enabling more important models (llama, qwen, etc.)

File: neural_speed/__init__.py (hunk 1 above)
Comment (zhentaoyu) [reply neural_speed/__init__.py:24]: Just an experimental value in SPR for gpt-j. Too large will affect `model mem scratch`. It will change if the use inputs bs is larger than `8`. No more ideas on how to set a more reasonable value. It may be changed if we have a better way. Any suggestions?

File: neural_speed/application/main_pybind.cpp (hunk 2 above)
Comment (zhentaoyu) [reply neural_speed/application/main_pybind.cpp:75]: @DDEle

File: neural_speed/application/main_pybind.cpp (hunk 3 above)
Comment (zhentaoyu) [reply neural_speed/application/main_pybind.cpp:291]: will do.

File: neural_speed/application/main_pybind.cpp (hunk 4 above)
Comment (zhentaoyu) [reply neural_speed/application/main_pybind.cpp:142]: These extra args are for `gpt_params`, as same as the `Model init`. So, pybind `gpt_params` and send it to related python apis? cc @zhenwei-intel

File: neural_speed/models/model_utils/model_types.h (hunk 5 above)
Comment (zhentaoyu) [reply neural_speed/models/model_utils/model_types.h:64]: Yes, should be.

File: neural_speed/models/model_utils/pool.cpp (hunk 6 above)
Comment (zhentaoyu) [reply neural_speed/models/model_utils/pool.cpp:40]: just for unified actions and check each part in the `step` process.

File: neural_speed/models/model_utils/scheduler.h (hunk 7 above)
Comment (zhentaoyu) [reply neural_speed/models/model_utils/scheduler.h:22]: `il` stands for `iteration-level`. Any suggested changes?

File: neural_speed/__init__.py (hunk 1 above)
Comment (airMeng) [reply neural_speed/__init__.py:24]: make it configurable (with a default value), we surely will not stick to SPR forever

File: neural_speed/models/model_utils/scheduler.h (hunk 7 above)
Comment (airMeng) [reply neural_speed/models/model_utils/scheduler.h:22]: ```Iter_level_worker_t``` shall be more explicit

Comment (zhentaoyu) [reply neural_speed/models/model_utils/scheduler.h:22]: how about `Iter_level_worker`? since other classes in `neural speed` do not have `_t` suffix.

File: neural_speed/application/main_pybind.cpp (hunk 2 above)
Comment (DDEle) [reply neural_speed/application/main_pybind.cpp:75]: The data structure happens to be the same. Currently, a query consists of a query ID and prompt tokens IDs, while a response consists of it responding query ID and generated token IDs. They may become different if we decide to add some fields to one but not to another.

File: neural_speed/models/model_utils/scheduler.h (hunk 7 above)
Comment (zhentaoyu) [reply neural_speed/models/model_utils/scheduler.h]: fixed in https://github.com/intel/neural-speed/pull/69/commits/949a463ef3412dcae2971989b678ffd3f62d3271

File: neural_speed/application/main_pybind.cpp (hunk 3 above)
Comment (zhentaoyu) [reply neural_speed/application/main_pybind.cpp]: added in https://github.com/intel/neural-speed/pull/69/commits/949a463ef3412dcae2971989b678ffd3f62d3271

File: neural_speed/application/main_pybind.cpp [hunk 8]
```diff
@@ -54,6 +56,252 @@
...
+  std::vector<Query> waiting;
+  // lock for waiting pool
+  std::mutex queue_mtx;
+  // status for telling server if it still need to continue running or not
+  // true: checking waiting pool and performing one step (or waiting new query)
+  // false: stop server
+  bool running;
+  gpt_params params;
+  // server policy (only FCFS (first come, first serve) now)
+  std::string policy;
```
Comment (airMeng) [inline neural_speed/application/main_pybind.cpp:294]: why not enum here?

File: neural_speed/application/main_pybind.cpp (hunk 2 above)
Comment (airMeng) [reply neural_speed/application/main_pybind.cpp:75]: > The data structure happens to be the same. Currently, a query consists of a query ID and prompt tokens IDs, while a response consists of it responding query ID and generated token IDs. They may become different if we decide to add some fields to one but not to another.

add this into comments

File: neural_speed/application/main_pybind.cpp (hunk 8 above)
Comment (zhentaoyu) [reply neural_speed/application/main_pybind.cpp:294]: added enum [here](https://github.com/intel/neural-speed/blob/cont_batching/neural_speed/models/model_utils/pool.h#L36-L39). And let the user pass a string may be easier than pass a pybind enum class?

File: neural_speed/application/main_pybind.cpp (hunk 2 above)
Comment (DDEle) [reply neural_speed/application/main_pybind.cpp:76]: Added in 664630b

**Review by zhenwei-intel — APPROVED** [review]
